### PR TITLE
Add SavePlanner and LoadPlanner support in Trainer

### PR DESCRIPTION
**Pull request by b-chu**
# What does this PR do?
This PR adds SavePlanner and LoadPlanner from `torch.distributed.checkpoint.planner` into Composer. These parameters are passed in when initializing trainer and are eventually passed to `save_checkpoint` and `load_sharded_checkpoint`. Because planners were implemented in torch 2, there's no type hinting, but type checking happens before the planners are used.

Some minor formatting fixes are included in this PR.

cc: @vchiley

---

**Stack**:
- #2698
- #2697
- #2696 ⬅
- #2695

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/utils/checkpoint.py [hunk 1]
```diff
@@ -628,6 +641,40 @@ def filter_func(state_dict: Dict) -> None:
...
+
+    Raises:
+        ValueError: If ``load_planner`` is not a
+            :class:`~torch.distributed.checkpoint.planner.LoadPlanner`.
+    """
+    from torch.distributed.checkpoint.planner import LoadPlanner
+
+    if load_planner is not None and not isinstance(load_planner, LoadPlanner):
+        raise ValueError((f'load_planner {type(load_planner)} is not a '
+                          'torch.distributed.checkpoint.planner.LoadPlanner'))
```
Comment (mvpatel2000) [inline composer/utils/checkpoint.py:675]: Do we need this? can we just rely on torch to handle this?

File: composer/utils/checkpoint.py [hunk 2]
```diff
@@ -318,6 +318,11 @@ def load_sharded_checkpoint(
             f'Sharded checkpoint loading on >1 node requires torch version >= 2.0.1. You have torch version {torch.__version__}'
         )
 
+    if state.fsdp_config is None:
+        raise ValueError('Sharded checkpointing requires an FSDP config.')
```
Comment (mvpatel2000) [inline composer/utils/checkpoint.py:322]:
```suggestion
    if state.fsdp_config is None:
        raise ValueError('Loading a sharded checkpointing requires passing a FSDP config to Trainer.')
```
nit: more descriptive and grammar

File: composer/utils/checkpoint.py (hunk 1 above)
Comment (b-chu) [reply composer/utils/checkpoint.py:675]: I want to throw the error as soon as we can so we don't waste time going down the code path. Usually this gets taken care of by type hints, but we can't do this because we support torch 1.13.

File: composer/utils/checkpoint.py [hunk 3]
```diff
@@ -416,7 +421,11 @@ def read_data(self, plan: LoadPlan, planner: LoadPlanner):
                 # Call function to modify state_dict
                 ignore_keys(model_state_dict)
 
-            dist_cp.load_state_dict(model_state_dict, storage_reader)
+            dist_cp.load_state_dict(
+                state_dict=model_state_dict,
+                storage_reader=storage_reader,
+                planner=load_planner,
+            )
```
Comment (mvpatel2000) [inline composer/utils/checkpoint.py:428]: Just to confirm, this works on all versions of torch >=2? No gating, right? Test implies this but want to be sure

File: tests/trainer/test_fsdp_checkpoint.py [hunk 4]
```diff
@@ -890,3 +892,183 @@ def test_cleanup_sharded_checkpoints(
...
+@pytest.mark.parametrize('optimizer', ['adam', 'adamw'])
+@pytest.mark.parametrize('state_dict_type', ['sharded'])
+@pytest.mark.parametrize('precision', ['amp_bf16', 'amp_fp16'])
+@pytest.mark.parametrize('use_remote', [pytest.param(True, marks=pytest.mark.remote), False])
+@pytest.mark.parametrize('autoresume', [True, False])
+@pytest.mark.parametrize('planner', [None, 'rename'])
+@pytest.mark.skipif(version.parse(torch.__version__) < version.parse('2.0'), reason='requires PyTorch 2.0 or higher')
+@pytest.mark.filterwarnings(r'ignore:TypedStorage is deprecated.:UserWarning')
+@pytest.mark.filterwarnings(r'ignore:.*metrics are not saved with sharded state dict.*:UserWarning')
+def test_fsdp_planner(
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:909]: Can you simplify this a bit since I don't think we need such a large matrix? I don't think a lot of these actually interact with loaders

In particular, I would remove:
- optimizer (adamw)
- precision (just bf16)
- use_remote (False)
- autoresume (False)
- dont need a test where planner = None?

and only keep weights only?

**Review by mvpatel2000 — COMMENTED** [review]
Mostly LGTM, one last round

File: tests/trainer/test_fsdp_checkpoint.py (hunk 4 above)
Comment (b-chu) [reply tests/trainer/test_fsdp_checkpoint.py:909]: Isn't it better to test more combinations?

The ones I think we definitely need are:
Planner = None vs not is useful because we don't want to make changes where None breaks.
I've had to make implementation changes for keep weights only vs not

File: composer/utils/checkpoint.py (hunk 3 above)
Comment (b-chu) [reply composer/utils/checkpoint.py:428]: Yes

File: tests/trainer/test_fsdp_checkpoint.py [hunk 5]
```diff
@@ -890,3 +892,183 @@ def test_cleanup_sharded_checkpoints(
...
+
+@pytest.mark.gpu
+@world_size(2)
+@pytest.mark.parametrize('weights_only', [False, True])
+@pytest.mark.parametrize('optimizer', ['adamw'])
+@pytest.mark.parametrize('state_dict_type', ['sharded'])
+@pytest.mark.parametrize('precision', ['amp_bf16'])
+@pytest.mark.parametrize('use_remote', [False])
+@pytest.mark.parametrize('autoresume', [False])
+@pytest.mark.parametrize('planner', [None, 'rename'])
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:905]: Can you flatten 1 choice options?

File: tests/trainer/test_fsdp_checkpoint.py [hunk 6]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+    s3_ephemeral_prefix,
+    request,
+):
+    optimizer = 'adamw'
+    state_dict_type = 'sharded'
+    precision = 'amp_bf16'
+    use_remote = False
+    autoresume = False
+    if weights_only and autoresume:
+        pytest.xfail('Weights only with autoresume is not supported')
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:919]:
```suggestion
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 7]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+    load_planner = planner
+    if planner == 'rename':
+        save_planner = RenameSavePlanner()
+        load_planner = RenameLoadPlanner()
+
+    if autoresume:
+        local_run_name = f'my-cool-autoresume-run-{uuid.uuid1()}'
+        run_name = dist.all_gather_object(local_run_name)[0]
+    else:
+        run_name = None
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:976]:
```suggestion
    run_name = None
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 8]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+    rng1 = get_rng_state()
+    state_dict_from_trainer1_ba2 = trainer1.state.state_dict()
+    trainer1.close()
+
+    if use_remote:
+        load_path = 's3://' + save_folder.strip('s3://').format(run_name=run_name) + '/ba2'
+        object_store = S3ObjectStore(bucket=f'{s3_bucket}')
+    else:
+        object_store = None
+        load_path = str(save_folder.format(run_name=run_name) / pathlib.Path('ba2'))
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:1015]:
```suggestion
    object_store = None
    load_path = str(save_folder.format(run_name=run_name) / pathlib.Path('ba2'))
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 9]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+        local_run_name = f'my-cool-autoresume-run-{uuid.uuid1()}'
+        run_name = dist.all_gather_object(local_run_name)[0]
+    else:
+        run_name = None
+
+    if use_remote:
+        save_folder = f's3://{s3_bucket}/{s3_ephemeral_prefix}/checkpoints/{{run_name}}'
+    else:
+        tmp_paths = dist.all_gather_object(os.path.abspath(tmp_path))
+        save_folder = os.path.join(tmp_paths[0], 'checkpoints', '{run_name}')
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:982]:
```suggestion
    tmp_paths = dist.all_gather_object(os.path.abspath(tmp_path))
    save_folder = os.path.join(tmp_paths[0], 'checkpoints', '{run_name}')
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 10]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+        assert is_checkpoint_legacy_sharded(
+            object_store=object_store,
+            source_path=load_path.replace(f's3://{s3_bucket}/', ''),
+        )
+    else:
+        assert not is_checkpoint_legacy_sharded(
+            object_store=object_store,
+            source_path=load_path.replace(f's3://{s3_bucket}/', ''),
+        )
+
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:1030]:
```suggestion
    assert not is_checkpoint_legacy_sharded(
        object_store=object_store,
        source_path=load_path.replace(f's3://{s3_bucket}/', ''),
    )

```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 11]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+    if autoresume:
+        load_path = None
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:1032]:
```suggestion
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 12]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+    trainer2 = get_trainer(
+        save_folder=str(save_folder),
+        save_filename=save_filename,
+        load_path=load_path,
+        precision=precision,
+        autoresume=autoresume,
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:1038]:
```suggestion
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 13]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+        load_path += '/' + load_filename
+        assert is_checkpoint_legacy_sharded(
+            object_store=object_store,
+            source_path=load_path.replace(f's3://{s3_bucket}/', ''),
+        )
+    else:
+        assert not is_checkpoint_legacy_sharded(
+            object_store=object_store,
+            source_path=load_path.replace(f's3://{s3_bucket}/', ''),
+        )
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:1029]:
```suggestion
    assert not is_checkpoint_legacy_sharded(
        object_store=object_store,
        source_path=load_path.replace(f's3://{s3_bucket}/', ''),
    )
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 14]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+    planner: Optional[str],
+    s3_bucket,
+    s3_ephemeral_prefix,
+    request,
+):
+    optimizer = 'adamw'
+    state_dict_type = 'sharded'
+    precision = 'amp_bf16'
+    use_remote = False
+    autoresume = False
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:917]:
```suggestion
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 15]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+    weights_only: bool,
+    planner: Optional[str],
+    s3_bucket,
+    s3_ephemeral_prefix,
+    request,
+):
+    optimizer = 'adamw'
+    state_dict_type = 'sharded'
+    precision = 'amp_bf16'
+    use_remote = False
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:916]:
```suggestion
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 16]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+    tmp_path: pathlib.Path,
+    weights_only: bool,
+    planner: Optional[str],
+    s3_bucket,
+    s3_ephemeral_prefix,
+    request,
+):
+    optimizer = 'adamw'
+    state_dict_type = 'sharded'
+    precision = 'amp_bf16'
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:915]:
```suggestion
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 17]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+    world_size,
+    tmp_path: pathlib.Path,
+    weights_only: bool,
+    planner: Optional[str],
+    s3_bucket,
+    s3_ephemeral_prefix,
+    request,
+):
+    optimizer = 'adamw'
+    state_dict_type = 'sharded'
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:914]:
```suggestion
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 18]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+            source_path=load_path.replace(f's3://{s3_bucket}/', ''),
+        )
+
+    if autoresume:
+        load_path = None
+    trainer2 = get_trainer(
+        save_folder=str(save_folder),
+        save_filename=save_filename,
+        load_path=load_path,
+        precision=precision,
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:1037]:
```suggestion
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 19]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+        autoresume=autoresume,
+        run_name=run_name,
+        max_duration='4ba',
+        save_interval='4ba',
+        optimizer=optimizer,
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:1042]:
```suggestion
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 20]
```diff
@@ -890,3 +892,178 @@ def test_cleanup_sharded_checkpoints(
...
+    if use_remote:
+        save_folder = f's3://{s3_bucket}/{s3_ephemeral_prefix}/checkpoints/{{run_name}}'
+    else:
+        tmp_paths = dist.all_gather_object(os.path.abspath(tmp_path))
+        save_folder = os.path.join(tmp_paths[0], 'checkpoints', '{run_name}')
+
+    save_filename = 'ba{batch}-rank{rank}.pt'
+
+    fsdp_config = FSDPConfig(
+        state_dict_type=state_dict_type,
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:987]:
```suggestion
        state_dict_type='sharded',
```

**Review by mvpatel2000 — COMMENTED** [review]
Can you please remove the code paths that are not used?

File: tests/trainer/test_fsdp_checkpoint.py [hunk 21]
```diff
@@ -890,3 +892,143 @@ def test_cleanup_sharded_checkpoints(
...
+@pytest.mark.filterwarnings(r'ignore:.*metrics are not saved with sharded state dict.*:UserWarning')
+def test_fsdp_planner(
+    world_size,
+    tmp_path: pathlib.Path,
+    weights_only: bool,
+    planner: Optional[str],
+    s3_bucket,
+    s3_ephemeral_prefix,
+    request,
+):
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:912]:
```suggestion
def test_fsdp_planner(
    world_size,
    tmp_path: pathlib.Path,
    weights_only: bool,
    planner: Optional[str],
):
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 22]
```diff
@@ -890,3 +892,143 @@ def test_cleanup_sharded_checkpoints(
...
+
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:913]:
```suggestion
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 23]
```diff
@@ -890,3 +892,143 @@ def test_cleanup_sharded_checkpoints(
...
+            self.is_coordinator = is_coordinator
+
+    save_planner = planner
+    load_planner = planner
+    if planner == 'rename':
+        save_planner = RenameSavePlanner()
+        load_planner = RenameLoadPlanner()
+
+    run_name = None
+
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:966]:
```suggestion

```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 24]
```diff
@@ -890,3 +892,143 @@ def test_cleanup_sharded_checkpoints(
...
+    tmp_paths = dist.all_gather_object(os.path.abspath(tmp_path))
+    save_folder = os.path.join(tmp_paths[0], 'checkpoints', '{run_name}')
+
+    save_filename = 'ba{batch}-rank{rank}.pt'
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:970]:
```suggestion
```

File: tests/trainer/test_fsdp_checkpoint.py [hunk 25]
```diff
@@ -890,3 +892,143 @@ def test_cleanup_sharded_checkpoints(
...
+
+    trainer1 = get_trainer(
+        save_folder=str(save_folder),
+        save_filename=save_filename,
+        run_name=run_name,
+        max_duration='2ba',
+        save_interval='2ba',
+        save_weights_only=weights_only,
+        fsdp_config=fsdp_config,
+    )
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:986]:
```suggestion
    trainer1 = get_trainer(
        save_folder=str(save_folder),
        max_duration='2ba',
        save_interval='2ba',
        save_weights_only=weights_only,
        fsdp_config=fsdp_config,
    )
```
Removes args which are same value as defaults

File: tests/trainer/test_fsdp_checkpoint.py [hunk 26]
```diff
@@ -890,3 +892,143 @@ def test_cleanup_sharded_checkpoints(
...
+    state_dict_from_trainer1_ba2 = trainer1.state.state_dict()
+    trainer1.close()
+
+    object_store = None
+    load_path = str(save_folder.format(run_name=run_name) / pathlib.Path('ba2'))
+
+    assert not is_checkpoint_legacy_sharded(
+        object_store=object_store,
+        source_path=load_path.replace(f's3://{s3_bucket}/', ''),
+    )
```
Comment (mvpatel2000) [inline tests/trainer/test_fsdp_checkpoint.py:999]:
```suggestion
    load_path = str(save_folder.format(run_name=run_name) / pathlib.Path('ba2'))

    assert not is_checkpoint_legacy_sharded(
        object_store=None,
        source_path=load_path,
    )
```

**Review by mvpatel2000 — APPROVED** [review]
A few more cleanups

File: tests/trainer/test_fsdp_checkpoint.py (hunk 22 above)
Comment (b-chu) [reply tests/trainer/test_fsdp_checkpoint.py:913]: This breaks lint

**Comment by b-chu** [comment]
✓ Commit merged in pull request [#2697](https://github.com/mosaicml/composer/pull/2697)
